File: C/xfactors.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

/* return the positive divisors ("factors") of n in ascending order.
   - n must be positive
   - out gets a malloc'd array that the caller must free
   - return value is the number of factors (0 on error)
*/
int factors(int n, int **out)
{
	int r, i;
	int cnt_small = 0, cnt_total = 0, k = 0;
	int *a;

	if (out == NULL) return 0;
	*out = NULL;

	if (n <= 0) return 0;

	r = (int)floor(sqrt((double)n));

	/* count small divisors (<= sqrt(n)) and total divisors */
	for (i = 1; i <= r; i++) {
		if (n % i == 0) {
			cnt_small++;
			if (i == n / i) cnt_total += 1;  /* perfect square */
			else            cnt_total += 2;
		}
	}

	a = (int *)malloc((size_t)cnt_total * sizeof(int));
	if (a == NULL) return 0;

	/* fill ascending:
	   1) small divisors increasing
	   2) corresponding large divisors decreasing (so overall increasing)
*/
	for (i = 1; i <= r; i++) {
		if (n % i == 0) a[k++] = i;
	}
	for (i = r; i >= 1; i--) {
		if (n % i == 0) {
			int j = n / i;
			if (j != i) a[k++] = j;
		}
	}

	*out = a;
	return cnt_total;
}

int main(void)
{
	int n;
	for (n = 1; n <= 10; n++) {
		int *f = NULL;
		int nf = factors(n, &f);

		printf("%d:", n);
		for (int i = 0; i < nf; i++) {
			printf(" %d", f[i]);
		}
		printf("\n");

		free(f);
	}
	return 0;
}

File: C/xfactors.f90
module xc2f_mod
implicit none
private
public :: factors
contains

pure function factors(n) result(out)
! return the positive divisors ("factors") of n in ascending order.
! - n must be positive
! - return value is the factors array (size(...) gives count; empty on error)
use, intrinsic :: iso_fortran_env, only: real64
integer, intent(in) :: n
integer, allocatable :: out(:)
integer :: cnt_small, cnt_total, i, j, k, r
if (n <= 0) then
   allocate(out(0))
   return
end if
cnt_small = 0
cnt_total = 0
k = 0
r = floor(sqrt(real(n, kind=real64)))
do i = 1, r
   if (mod(n, i) == 0) then
      cnt_small = cnt_small + 1
      ! count small divisors (<= sqrt(n)) and total divisors
      cnt_total = cnt_total + merge(1, 2, i == n / i)
   end if
end do
! perfect square
allocate(out(cnt_total))
do i = 1, r
   if (mod(n, i) == 0) then
      out(k+1) = i
      k = k + 1
   end if
end do
do i = r, 1, -1
   if (mod(n, i) == 0) then
      j = n / i
      ! fill ascending:
      ! 1) small divisors increasing
      ! 2) corresponding large divisors decreasing (so overall increasing)
      if (j /= i) then
         out(k+1) = j
         k = k + 1
      end if
   end if
end do
end function factors

end module xc2f_mod

program main
use xc2f_mod, only: factors
implicit none
integer :: n
do n = 1, 10
   write(*,"(i0,a,*(1x,i0))") n, ":", factors(n)
end do
end program main
